src/main/java/net/minecraftforge/common/util/MutableHashedLinkedList.java

```diff
@@ -0,0 +1,205 @@
+        return null;
+    }
+
+    public boolean contains(K key) { return this.entries.containsKey(key); }
+    public boolean isEmpty() { return this.entries.isEmpty(); }
+
+    @Nullable
+    public V remove(K key) {
+        var ret = this.entries.remove(key);
+        if (ret == null)
+            return null;
+
+        ret.remove();
+        return ret.getValue();
+    }
+
+    @Nullable
+    public V get(K key) {
+        var entry = entries.get(key);
+        return entry == null ? null : entry.getValue();
+    }
+
+    @Override
+    public Iterator<Map.Entry<K, V>> iterator() {
+        return new Iterator<Map.Entry<K, V>>() {
+            private Entry current = head.getValue();
+
+            @Override
+            public boolean hasNext() {
+                return current != null;
+            }
+
+            @Override
+            public Map.Entry<K, V> next() {
+                if (!hasNext())
+                    throw new NoSuchElementException();
+                var ret = current;
+                current = current.next;
+                return ret;
+            }
+        };
+    }
+
+    @Nullable
+    public V putFirst(K key, V value) {
+        if (head.getValue() != null)
+            return putBefore(head.getValue().getKey(), key, value);
+        return put(key, value);
+    }
+
+    @Nullable
+    public V putAfter(K after, K key, V value) {
+        var target = entries.get(after);
+        if (target == null)
+            return put(key, value);
+
+        V ret = null;
+        var entry = entries.get(key);
+        if (entry != null) {
+            ret = entry.value;
+            entry.value = merge.apply(key, ret, value);
+            entry.remove();
+        } else {
+            entry = new Entry(key, value);
+            entries.put(key, entry);
+        }
+
+        entry.previous = target;
+        if (target.next == null)
+            last.setValue(target);
+        else
+            target.next.previous = entry;
+        entry.next = target.next;
+
+        target.next = entry;
+        return ret;
+    }
+
+    @Nullable
+    public V putBefore(K before, K key, V value) {
+        var target = entries.get(before);
+        if (target == null)
+            return put(key, value);
+
+        V ret = null;
+        var entry = entries.get(key);
+        if (entry != null) {
+            ret = entry.value;
+            entry.value = merge.apply(key, ret, value);
+            entry.remove();
+        } else {
+            entry = new Entry(key, value);
+            entries.put(key, entry);
+        }
+
+        entry.previous = target.previous;
+        if (target.previous == null)
+            head.setValue(entry);
+        else
+            target.previous.next = entry;
+        entry.next = target;
+
+        target.previous = entry;
+
+        return ret;
+    }
+
+    public static interface MergeFunction<Key, Value> {
+        Value apply(Key key, Value left, Value right);
```

Comment:
The use of identifiers for the type parameters that are very similar to plausible class names unnerves me a bit. Perhaps they should be just the usual `K` and `V`? Though this is only a personal preference on my part.

Reply:
If I do not use single letters, I tend to differentiate generic types by making them all caps. Same case as a constant, but a constant will never be a type so no ambiguity there.

Reply:
Meh, I needed something that wasn't `K` and '`V` because its a static interface and cant share the outer generics {javac yelled at me}
It doesn't *really* matter as you're most likely gunna just use a lambda in the constructor. Also, it's a generic so we could change it to whatever and not be a breaking change.